content/hardware/03.nano/boards/nano-r4/tutorials/04.external-interrupts/content.md

```diff
@@ -0,0 +1,478 @@
+2. **Separate ISRs and flags**: Each interrupt has its own ISR (`button1ISR` and `button2ISR`) and its own flag variable. This keeps the interrupt handling clean and prevents cross-contamination between events.
+3. **RGB LED control**: The Nano R4's RGB LEDs are active-low, meaning `LOW` turns them on and `HIGH` turns them off. This is opposite to the built-in orange LED, which is why we initialize them to `HIGH` in setup.
+4. **Counter variables**: Each ISR increments its own counter, demonstrating that ISRs can safely modify multiple volatile variables. These counters help verify that no interrupts are being missed, even when both buttons are pressed rapidly.
+
+## Best Practices 
+
+### ISR Design Rules
+
+Writing effective Interrupt Service Routines requires discipline. These guidelines will help you avoid the most common pitfalls:
+
+- **Keep ISRs extremely short**. Your ISR should execute in microseconds, not milliseconds. The golden rule is to do the absolute minimum work necessary (typically just setting a flag or incrementing a counter). Complex operations belong in your main loop where they do not block other interrupts.
+- **Always use `volatile` for shared variables**. Any variable modified in an ISR and read in your main code (or vice versa) must be declared `volatile`. This prevents the compiler from caching the variable in a register where changes are not be visible between the ISR and main code.
+- **Never call blocking functions**: Functions like `delay()`, `Serial.print()`, `millis()`, or anything that waits for hardware are forbidden in ISRs. These functions often rely on interrupts themselves and calling them from an ISR can cause deadlocks or crashes.
+- **Protect multi-byte shared data**. When your main code needs to read a multi-byte variable that an ISR modifies (like `long` or `float`), you must temporarily disable interrupts to prevent the value from changing mid-read. 
+
+Here is how to safely read a multi-byte variable that is modified by an ISR:
+
+```arduino
+// Reading a 4-byte counter safely
+volatile unsigned long pulseCount = 0;
+
+void countISR() {
+  pulseCount++;  // ISR modifies this
+}
+
+void loop() {
+  // Wrong way - could read corrupted value
+  unsigned long count = pulseCount;  // Dangerous!
+  
+  // Right way - atomic read
+  noInterrupts();
+  unsigned long count = pulseCount;  // Safe
+  interrupts();
+  
+  // Now use 'count' for calculations
+}
+```
+### Common Mistakes to Avoid
+
+Even experienced developers fall into these traps when working with interrupts. Understanding these common mistakes will save you hours of debugging:
+
+- **Channel conflicts on the Nano R4 board**: The most frustrating error is when interrupts mysteriously stop working because you have accidentally used two pins on the same channel.The second attachment silently overwrites the first with no warning:

```

Comment:
```suggestion
- **Channel conflicts on the Nano R4 board**: The most frustrating error is when interrupts mysteriously stop working because you have accidentally used two pins on the same channel. The second attachment silently overwrites the first with no warning:

```
Minor suggestion